### PR TITLE
OSD/ReplicatedPG: Fixes unneccessary object promotion when deleting f…

### PR DESCRIPTION
**Pull request by ifed01**
…rom the cache that lacks hit_set configured.

Minor cleanup nearby too.

This issue was mentioned at http://tracker.ceph.com/issues/13894 ( comment #12 and beyond ) but it's probably not the root cause of 13894.

Signed-off-by: Igor Fedotov <ifedotov@mirantis.com>

### CODE REVIEW

File: src/osd/ReplicatedPG.cc [hunk 1]
```diff
@@ -2160,11 +2160,12 @@ ReplicatedPG::cache_result_t ReplicatedPG::maybe_handle_cache_detail(
...
       // Promote too?
-      if (!op->need_skip_promote()) {
-        maybe_promote(obc, missing_oid, oloc, in_hit_set,
+      if (!op->need_skip_promote() && 
+          maybe_promote(obc, missing_oid, oloc, in_hit_set,
 	              pool.info.min_write_recency_for_promote,
 		      OpRequestRef(),
-		      promote_obc);
+		      promote_obc)) {
+	return cache_result_t::BLOCKED_PROMOTE;
```
Comment (liewegas) [inline src/osd/ReplicatedPG.cc]: This hunk looks like an unrelated bug fix.. can you move it into a separate patch?

File: src/osd/ReplicatedPG.cc [hunk 2]
```diff
@@ -2274,7 +2275,7 @@ bool ReplicatedPG::maybe_promote(ObjectContextRef obc,
 				 ObjectContextRef *promote_obc)
 {
   dout(20) << __func__ << " missing_oid " << missing_oid
-	   << "  in_hit_set " << in_hit_set << dendl;
+	   << "  in_hit_set " << in_hit_set << " recency "<<recency<<dendl;
```
Comment (liewegas) [inline src/osd/ReplicatedPG.cc]: whitespace

File: src/osd/ReplicatedPG.cc [hunk 3]
```diff
@@ -12078,7 +12078,8 @@ bool ReplicatedPG::agent_choose_mode(bool restart, OpRequestRef op)
 
 void ReplicatedPG::agent_estimate_temp(const hobject_t& oid, int *temp)
 {
-  assert(hit_set);
+  if (!hit_set)
+    return;
   assert(temp);
   *temp = 0;
```
Comment (liewegas) [inline src/osd/ReplicatedPG.cc]: shoudl we set *temp to 0 before the return?

**Comment by liewegas** [comment]
can you also move the cleanup into a separate patch?  Thanks!

**Comment by ifed01** [comment]
@liewegas, fixed.
